### PR TITLE
Update to new sim protocol

### PR DESCRIPTION
**Pull request by kylestach**
## Description
This also includes some minor compatability fixes and reworks to match
the new simulator setup - for example, kick speeds are now specified in
m/s all the way through.

**Review note**: please do not review the `.proto` files; they are long and not ours.

## Associated Issue
[ClickUp](https://app.clickup.com/t/vtbc83)

### Regular sim run
1. Download ER-force's simulator: https://github.com/robotics-erlangen/framework
2. Build and run `simulator-cli` using their instructions (run the executable in framework/build/bin)
3. Comment out the `grsim` node in `sim.launch.py` and build
4. Launch sim

Expected result: Simulation runs correctly; placement works

### Old grSim run
1. Launch `sim.launch.py` without the change above

Expected result: the old grSim config should run

### CODE REVIEW

File: soccer/src/soccer/radio/sim_radio.cpp [hunk 1]
```diff
@@ -18,28 +18,40 @@ using namespace boost::asio;
...
     }
+
     if (!placement.ball.velocity.empty()) {
         const auto& velocity = placement.ball.velocity.at(0);
-        packet.mutable_replacement()->mutable_ball()->set_vx(velocity.x);
-        packet.mutable_replacement()->mutable_ball()->set_vy(velocity.y);
+        control->mutable_teleport_ball()->set_vx(velocity.x);
+        control->mutable_teleport_ball()->set_vy(velocity.y);
+        control->mutable_teleport_ball()->set_vz(0);
+        control->mutable_teleport_ball()->set_teleport_safely(true);
```
Comment (kylestach) [inline soccer/src/soccer/radio/sim_radio.cpp]:
```suggestion
```

File: rj_common/include/rj_common/utils.hpp [hunk 2]
```diff
@@ -26,6 +26,11 @@ inline T apply_low_pass_filter(const T& old_value, const T& new_value, double ga
     return gain * new_value + (1 - gain) * old_value;
 }
 
+template <typename T>
+inline T lerp(const T& a, const T& b, double factor) {
+    return factor * a + (1 - factor) * a;
```
Comment (kylestach) [inline rj_common/include/rj_common/utils.hpp]:
```suggestion
    return factor * a + (1 - factor) * b;
```

**Review by kfu02 — CHANGES_REQUESTED** [review]
Switching to ERForce works flawlessly, provided the `--realism None` flag is set for `./simulator-cli`.

Going back to grSim is a little rocky. After reverting the `sim.launch.py` edits, I ran the following:
```
rm -rf build-*
git clean -fX install/
make perf
ros2 launch rj_robocup sim.launch.py
```
which opened grSim again, but all ball/robot placement was broken, and the motion commands sent by gameplay didn't move any robots. 

I had to additionally call `./util/ubuntu-setup` before re-running all the above commands to get grSim working again. If it's an issue that can be fixed later without too much consequence in the meantime then ship it.

**Review by HussainGynai — CHANGES_REQUESTED** [review]
Same thing as Kevin pretty much, there are 5 extra bots on each team showing up (starting on the wrong side of the field) along the right side. Is it do with the Division Enum?

File: soccer/src/soccer/radio/packet_convert.cpp [hunk 3]
```diff
@@ -251,38 +239,36 @@ void to_proto(const RobotIntent& intent, const MotionSetpoint& setpoint, int she
...
         } else {
             // Chip kick
-            constexpr double kChipAngle = 40 * M_PI / 180;  // degrees
-
-            double speed = intent.kick_speed;
-            grsim->set_kickspeedx(static_cast<float>(std::cos(kChipAngle) * speed));
-            grsim->set_kickspeedz(static_cast<float>(std::sin(kChipAngle) * speed));
+            sim->set_kick_speed(intent.kick_speed);
+            // Chip angle is already in degrees
+            sim->set_kick_angle(PARAM_chip_angle);
```
Comment (kylestach) [inline soccer/src/soccer/radio/packet_convert.cpp]:
```suggestion
            sim->set_kick_angle(static_cast<float>(PARAM_chip_angle));
```

**Review by kfu02 — APPROVED** [review]
Approved bc the crashing issues are annoying, but not gamebreaking.

**Review by HussainGynai — APPROVED** [review]

File: soccer/src/soccer/radio/packet_convert.cpp [hunk 4]
```diff
@@ -251,38 +239,36 @@ void to_proto(const RobotIntent& intent, const MotionSetpoint& setpoint, int she
...
-    grsim->set_velnormal(-static_cast<float>(setpoint.xvelocity));
-    grsim->set_velangular(static_cast<float>(setpoint.avelocity));
+    auto* command = sim->mutable_move_command()->mutable_local_velocity();
+    command->set_forward(static_cast<float>(setpoint.yvelocity));
+    command->set_left(-static_cast<float>(setpoint.xvelocity));
+    command->set_angular(static_cast<float>(setpoint.avelocity));
 
-    grsim->set_spinner(intent.dribbler_speed > 0);
-    grsim->set_wheelsspeed(false);
+    sim->set_dribbler_speed(PARAM_max_dribbler_speed * intent.dribbler_speed);
```
Comment (kylestach) [inline soccer/src/soccer/radio/packet_convert.cpp]:
```suggestion
    sim->set_dribbler_speed(static_cast<float>(PARAM_max_dribbler_speed * intent.dribbler_speed));
```

File: soccer/src/soccer/radio/packet_convert.cpp [hunk 5]
```diff
@@ -297,45 +283,48 @@ void ros_to_rtp(const rj_msgs::msg::ManipulatorSetpoint& manipulator,
...
-    grsim->set_velnormal(-static_cast<float>(motion.velocity_x_mps));
-    grsim->set_velangular(static_cast<float>(motion.velocity_z_radps));
+    auto* command = sim->mutable_move_command()->mutable_local_velocity();
+    command->set_forward(static_cast<float>(motion.velocity_y_mps));
+    command->set_left(-static_cast<float>(motion.velocity_x_mps));
+    command->set_angular(static_cast<float>(motion.velocity_z_radps));
 
-    grsim->set_spinner(manipulator.dribbler_speed > 0);
-    grsim->set_wheelsspeed(false);
+    sim->set_dribbler_speed(PARAM_max_dribbler_speed * manipulator.dribbler_speed);
```
Comment (kylestach) [inline soccer/src/soccer/radio/packet_convert.cpp]:
```suggestion
    sim->set_dribbler_speed(static_cast<float>(PARAM_max_dribbler_speed * manipulator.dribbler_speed));
```

**Pull request merged**